### PR TITLE
Adjustments of loader-v4

### PR DESCRIPTION
**Pull request by Lichtso**
#### Problem
While writing the [SIMD for loader-v4](https://github.com/solana-foundation/solana-improvement-documents/pull/167) I noticed some minor issues with our implementation.

#### Summary of Changes
- Removes the empty program account check (`InvalidAccountData`), which is redundant because of the subsequent minimum size check.
- Adjusts checks of `program.state.status` to conform to error messages.
- Adds check to prevent authority transfer to itself.
- Adjusts most execution errors to `UnsupportedProgramId`.
- Removes the redundant owner check in execution (`InvalidAccountOwner`).
- Adds a tombstone in the program cache upon retraction.

### CODE REVIEW

File: programs/loader-v4/src/lib.rs [hunk 1]
```diff
@@ -488,12 +484,22 @@ pub fn process_instruction_retract(
         );
         return Err(InstructionError::InvalidArgument);
     }
-    if matches!(state.status, LoaderV4Status::Retracted) {
+    if !matches!(state.status, LoaderV4Status::Deployed) {
         ic_logger_msg!(log_collector, "Program is not deployed");
```
Comment (pgarg66) [inline programs/loader-v4/src/lib.rs:488]: The log message should also be changed, as the status can be `Finalized`.

File: programs/loader-v4/src/lib.rs [hunk 2]
```diff
@@ -575,26 +585,18 @@ pub fn process_instruction_inner(
         .map_err(|err| Box::new(err) as Box<dyn std::error::Error>)
     } else {
         let program = instruction_context.try_borrow_last_program_account(transaction_context)?;
-        if !loader_v4::check_id(program.get_owner()) {
```
Comment (pgarg66) [inline programs/loader-v4/src/lib.rs:578]: Could you explain why this check is not needed?

Comment (Lichtso) [reply programs/loader-v4/src/lib.rs:578]: The program runtime invokes programs by their owner (see https://github.com/anza-xyz/agave/blob/4daf93c3d37d92a7803eb30feece13f8d36a5834/program-runtime/src/invoke_context.rs#L482C28-L482C49), so this is implicitly always true here.

File: programs/loader-v4/src/lib.rs (hunk 1 above)
Comment (Lichtso) [reply programs/loader-v4/src/lib.rs:488]: The `check_program_account()` above prevents the `Finalized` case.

**Review by pgarg66 — APPROVED** [review]

**Pull request merged**